internal/core/domain/market_service.go

```diff
@@ -217,6 +219,176 @@ func (m *Market) ChangeQuotePrice(price decimal.Decimal) error {
 	return nil
 }
 
+func (m *Market) Preview(
+	baseBalance, quoteBalance, amount uint64,
+	isBaseAsset, isBuy bool,
+) (*PreviewInfo, error) {
+	if !m.IsTradable() {
+		return nil, ErrMarketIsClosed
+	}
+
+	if isBaseAsset {
+		if amount < uint64(m.FixedFee.BaseFee) {
+			return nil, ErrMarketPreviewAmountTooLow
+		}
+	} else {
+		if amount < uint64(m.FixedFee.QuoteFee) {
+			return nil, ErrMarketPreviewAmountTooLow
+		}
+	}
+
+	formula := m.formula(isBaseAsset, isBuy)
+	args := m.formulaOpts(baseBalance, quoteBalance, isBaseAsset, isBuy)
+
+	price, err := m.priceForStrategy(baseBalance, quoteBalance)
+	if err != nil {
+		return nil, err
+	}
+
+	previewAmount, err := formula(args, amount)
+	if err != nil {
+		return nil, err
+	}
+
+	previewAmount, err = m.chargeFixedFees(
+		baseBalance, quoteBalance, previewAmount, isBaseAsset, isBuy,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	previewAsset := m.BaseAsset
+	if isBaseAsset {
+		previewAsset = m.QuoteAsset
+	}
+
+	return &PreviewInfo{
+		Price:  price,
+		Amount: previewAmount,
+		Asset:  previewAsset,
+	}, nil
+}
+
+func (m *Market) formula(
+	isBaseAsset, isBuy bool,
+) func(interface{}, uint64) (uint64, error) {
+	formula := m.getStrategySafe().Formula()
+	if isBuy {
+		if isBaseAsset {
+			return formula.InGivenOut
+		}
+		return formula.OutGivenIn
+	}
+
+	if isBaseAsset {
+		return formula.OutGivenIn
+	}
+	return formula.InGivenOut
+}
+
+func (m *Market) formulaOpts(
```

Comment:
I would rather uplift the branch (if strategy or not) and create a dedicated method like `formulaOptsForPluggable`, `formulaOptsForBalanced` mostly beacuse we can this way add more formulas later, otherwise this methos will hide all the logic.